overlord/devicestate/devicestate_serial_test.go

```diff
@@ -2408,3 +2409,194 @@ func (s *deviceMgrSerialSuite) TestDeviceSerialRestoreHappy(c *C) {
 	c.Check(log.String(), testutil.Contains,
 		fmt.Sprintf("restored serial serial-1234 for my-brand/pc-20 signed with key %v", devKey.PublicKey().ID()))
 }
+
+func (s *deviceMgrSerialSuite) TestShouldRequestSerial(c *C) {
+	type testCase struct {
+		deviceServiceAccess string
+		deviceServiceURL    string
+		storeAccess         string
+		gadgetName          string
+		expected            bool
+	}
+
+	testCases := []testCase{
+		{
+			storeAccess:         "",
+			gadgetName:          "",
+			deviceServiceAccess: "",
+			deviceServiceURL:    "",
+			expected:            true,
+		},
+		{
+			storeAccess:         "offline",
+			gadgetName:          "",
+			deviceServiceAccess: "",
+			deviceServiceURL:    "",
+			expected:            false,
+		},
+		{
+			storeAccess:         "",
+			gadgetName:          "gadget",
+			deviceServiceAccess: "offline",
+			deviceServiceURL:    "",
+			expected:            false,
+		},
+		{
+			storeAccess:         "",
+			gadgetName:          "gadget",
+			deviceServiceAccess: "",
+			deviceServiceURL:    "",
+			expected:            true,
+		},
+		{
+			storeAccess:         "offline",
+			gadgetName:          "gadget",
+			deviceServiceAccess: "",
+			deviceServiceURL:    "https://example.com",
+			expected:            true,
+		},
+		{
+			storeAccess:         "offline",
+			gadgetName:          "gadget",
+			deviceServiceAccess: "",
+			deviceServiceURL:    "",
+			expected:            false,
+		},
+		{
+			storeAccess:         "offline",
+			gadgetName:          "gadget",
+			deviceServiceAccess: "",
+			deviceServiceURL:    "https://example.com",
+			expected:            true,
+		},
+	}
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	for _, t := range testCases {
+		tr := config.NewTransaction(s.state)
+		err := tr.Set("core", "store.access", t.storeAccess)
+		c.Assert(err, IsNil)
+
+		if t.gadgetName != "" {
+			err = tr.Set(t.gadgetName, "device-service.access", t.deviceServiceAccess)
+			c.Assert(err, IsNil)
+			err = tr.Set(t.gadgetName, "device-service.url", t.deviceServiceURL)
+			c.Assert(err, IsNil)
+		}
+		tr.Commit()
+
+		shouldReqest, err := devicestate.ShouldRequestSerial(s.state, t.gadgetName)
+		c.Check(err, IsNil)
+		c.Check(shouldReqest, Equals, t.expected, Commentf("%+v", t))
+	}
+}
+
+func (s *deviceMgrSerialSuite) TestDeviceManagerFullAccess(c *C) {
+	s.state.Lock()
+	devicestatetest.SetDevice(s.state, &auth.DeviceState{
+		Brand: "canonical",
+		Model: "pc",
+	})
+
+	s.makeModelAssertionInState(c, "canonical", "pc", map[string]interface{}{
+		"architecture": "amd64",
+		"kernel":       "pc-kernel",
+		"gadget":       "pc",
+	})
+
+	devicestatetest.MockGadget(c, s.state, "pc", snap.R(2), nil)
+	s.state.Set("seeded", true)
+	s.state.Unlock()
+
+	err := s.mgr.Ensure()
+	c.Assert(err, IsNil)
+
+	s.state.Lock()
+	change := s.findBecomeOperationalChange()
+	tasks := change.Tasks()
+	s.state.Unlock()
+
+	sort.Slice(tasks, func(l, r int) bool {
```

Comment:
There's a `testutil.DeepUnsortedMatches` that can compare these things without having to sort them

Reply:
Thanks, that is useful to know! In this case though, I'd just need to verify the kind of task.